wpengine/atlas-content-modeler
# [MTKA-1395] Import ACM Blueprints via WP-CLI

Adds the ability to import an ACM blueprint using WP-CLI.

https://wpengine.atlassian.net/browse/MTKA-1395

## Testing

Includes integration tests for import functions.

### To test manually

Delete existing ACM models to prevent import conflicts on your first test run.

1. Open a shell ready for WP-CLI (in Local, right-click the site and choose “Open Site Shell”).
2. Run `wp acm blueprint import https://contentmodel1.wpengine.com/wp-content/uploads/2022/02/acm-rabbits.zip`.

Your WP instance should now contain a rabbit model, related taxonomy, media, and rabbit entries with pre-filled data and images.

To reset state for another test (excludes removing posts and media):

```
wp term delete breed cartoon --by=slug && wp term delete breed human --by=slug && wp option delete atlas_content_modeler_taxonomies && wp option delete atlas_content_modeler_post_types
```

To reset only model state (will result in warnings about duplicate taxonomies if importing the same zip file on the next run):

```
wp option delete atlas_content_modeler_post_types
```

To test a blueprint that will fail version checking:

```
wp acm blueprint import http://contentmodel1.wpengine.com/wp-content/uploads/2022/02/acm-needs-newer-version.zip
```

## Screenshots


Successful run (exit 0):

```
> wp acm blueprint import https://contentmodel1.wpengine.com/wp-content/uploads/2022/02/acm-rabbits.zip
Fetching zip.
Unzipping.
Verifying ACM manifest.
Checking minimum versions.
Importing ACM models and fields.
Importing ACM taxonomies.
Importing posts.
Importing terms.
Tagging posts.
Importing media.
Importing post meta.
Restoring ACM relationships.
Deleting zip and manifest.
Success: Import complete.
```

Failed run (fails version check, exit 1):

```
> wp acm blueprint import http://contentmodel1.wpengine.com/wp-content/uploads/2022/02/acm-needs-newer-version.zip
Fetching zip.
Unzipping.
Verifying ACM manifest.
Checking minimum versions.
Error: acm.json requires an ACM version of 100000000000000000000 but the current ACM version is 0.13.0.
```

Failed run (model conflict, exit 1):

```
> wp acm blueprint import https://contentmodel1.wpengine.com/wp-content/uploads/2022/02/acm-rabbits.zip
Cannot load Zend OPcache - it was already loaded
Fetching zip.
Unzipping.
Verifying ACM manifest.
Checking minimum versions.
Importing ACM models and fields.
Error: A model with slug ‘rabbit’ already exists.
```

Successful run with warnings (exit 0):

```
> wp acm blueprint import https://contentmodel1.wpengine.com/wp-content/uploads/2022/02/acm-rabbits.zip
Fetching zip.
Unzipping.
Verifying ACM manifest.
Checking minimum versions.
Importing ACM models and fields.
Importing ACM taxonomies.
Warning: Errors encountered during taxonomy import.
Warning: A taxonomy with an ID of ‘breed’ already exists.
Importing posts.
Importing terms.
Warning: Errors encountered during term import.
Warning: A term with the name provided already exists in this taxonomy.
Warning: A term with the name provided already exists in this taxonomy.
Tagging posts.
Importing media.
Importing post meta.
Restoring ACM relationships.
Deleting zip and manifest.
Success: Import complete.
```

(Note that the vague “term with the name provided” [comes from WP core](https://github.com/WordPress/WordPress/blob/db603605ac3465185c6b07f9b0351f325c28f0e6/wp-includes/taxonomy.php#L2477), not ACM.)

## Documentation Changes

Will document separately once the export command exits.

## Dependant PRs

n/a
